### PR TITLE
Update build-related instructions #8271

### LINKED ISSUE #8115: Dev docs: mention when to use GodMode
Currently for a new contributor approaching this project, there is no mention about why someone would want to use [GodMode](https://github.com/TEAMMATES/teammates/blob/master/docs/godmode.md) or how to use it in the Development Workflow documentation or other introductory documentation. 

The one mention I could find was in [Development Guidelines Page](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md) and even then it just has a link to the GodMode page and nothing about when someone would need to use it. 

So the task would be to maybe add a small comment explaining the basic idea of GodMode and why someone would want to use it along with a link to the [GodMode page](https://github.com/TEAMMATES/teammates/blob/master/docs/godmode.md) in the [Development Workflow page](https://github.com/TEAMMATES/teammates/blob/master/docs/process.md) or maybe another page that's suitable or expanding on the mention of it in the [Development Guidelines Page](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md)

### LINKED ISSUE #8173: static-analysis.md: fix CheckStyle and PMD Eclipse plugin link
**Environment**

Tested in : Firefox 46.0<br>Chrome Version 62.0.3202.94

**Steps to reproduce**
While installing the plugins from the [static-analysis](https://github.com/TEAMMATES/teammates/blob/master/docs/static-analysis.md) file, I noticed that the download link in `Configuring PMD Eclipse plugin` under `PMD` gives `403 Forbidden` error.
screenshot 249

**Expected behaviour**
This link should lead to a site that is accessible by everyone.

**Actual behaviour**
screenshot 248

### LINKED ISSUE #8177: AppVeyor: Run all the lint routines
When AppVeyor was set up back then (#6238), the focus was to get it up and running first, and the lint routines were all excluded. Since then, issues such as #7020 and #7469 had surfaced, which could be prevented if all the lint routines are also run in AppVeyor CI.
The loss in performance in doing so is marginal to zero since AppVeyor is only run on `master` and `release` branches.

### LINKED ISSUE #8178: Dev docs: mention the need for npm run build
Hello TEAMMATES,

While trying to setup TEAMMATES in our local machines we followed the steps provided in the "Development Guidelines" of TEAMMATES documentation. 

After the installation process was done in the dev server, while following the guidelines for "Logging in to a TEAMMATES instance", we tried adding a new instructor in the adminhome page. However it did not work as the submit button was not enabled. Also we could not locate any instructions in the documentation on how to resolve it. 
So we researched the problem and came up with the solution to add “**npm run build**” first and then try adding the instructor to the dev server from the Admin Home page. This detail needs to be added to the "**Development Guidelines**" of the TEAMMATES project documentation under “As Instructor” after the instructor enters his/her credentials so that new users don’t face any issue while testing for adding new instructors in the dev server.

Thanks
Team Mavericks(Niluja Biswal, Neeharica Kodali, Durgabhavani Peta, Kasi Viswanath)
University Of Nebraska,Omaha

### LINKED ISSUE #8219: ide-usage.md: Eclipse setup: update text encoding instruction
As discovered by @sukanta-27 in https://github.com/TEAMMATES/teammates/issues/8196#issuecomment-350508266, setting the project-specific text encoding is insufficient. Update the instructions in `ide-usage.md` to indicate that the workspace-wide preference should be set instead.

### LINKED ISSUE #8238: static-analysis.md: fix broken PMD documentation link
**Steps to reproduce**
1. Proceed to static-analysis.md.
2. Go to `Tool Stack` > `PMD` > `Suppressing PMD warnings` and click on the link indicated by "...documentation here".

**Expected behaviour**

The link should lead to further documentation that explains how PMD provides several other methods of suppressing rule violations. 

**Actual behaviour**

Instead, a 404 error was displayed after clicking on the link:

screen shot 2017-12-22 at 2 18 32 am

### LINKED ISSUE #8260: Add custom gradle task for deployment
Edit by @whipermr5: Write a custom gradle task that encompasses `./gradlew appengineDeploy appengineDeployCron appengineDeployIndex appengineDeployQueue` and update the instructions in `development.md`.

<hr>

When trying to log in as instructor in my app teammates-shradheya.appspot.com

I am getting the following error after I add an instructor and attempt to join as instructor using the unique link mentioned in the email sent and then logging in using the google account( Note: I have used a different Google account than the one i used to set up the GCP, it is the account on which I got the email of the instructor joining link)

34327113-58d42152-e8e2-11e7-9f9d-02a080cdb04b

I am not sure what could be the reason as to why I am not able to access the instructor page.

### PR DESCRIPTION
**Pull request by wkurniawan07**
- [x] Fixes #8178 
- [x] Fixes #8177 
- [x] Fixes #8260 
- [x] Fixes #8173 
  - Here the CheckStyle version has to be precisely 8.0.0 due to #8278
- [x] Fixes #8238 
- [x] Fixes #8219 
- [x] Fixes #8115 

This PR is to be rebased-and-merged.

### CODE REVIEW

**Comment by LiHaoTan** [comment]
@wkurniawan07 Can we use `deployables` instead?

```Gradle
    deploy {
        String appengineWebXmlPath = "${projectDir}/src/main/webapp/WEB-INF/appengine-web.xml"
        Node appengineWebXml = new File(appengineWebXmlPath).exists() ? new XmlParser().parse(appengineWebXmlPath) : null
        project = appengineWebXml == null ? null : appengineWebXml.application.text()
        version = appengineWebXml == null ? null : appengineWebXml.version.text()
        stopPreviousVersion = false
        promote = false
        deployables = ["${appEngineDirectory}/app.yaml",
                       "${appEngineDirectory}/cron.yaml",
                       "${appEngineDirectory}/index.yaml",
                       "${appEngineDirectory}/queue.yaml"]
    }
```

**Comment by wkurniawan07** [comment]
@LiHaoTan if I do that, how do I deploy just the app and nothing else?

**Comment by whipermr5** [comment]
@wkurniawan07 Arguably, the old gradle plugin didn't allow us to do that.

**Comment by wkurniawan07** [comment]
@whipermr5 and now that the new plugin actually allows us to, why waste the feature, right?

**Comment by LiHaoTan** [comment]
@wkurniawan07 I'm not sure but I believe that there is usually no need to just deploy the app, can't think of a use case for it. My belief is that because they are consolidating everything into Google Cloud SDK, they made it more flexible to support everything together.

On the other hand, if we do need to have multiple configurations like for deploying multiple services separately, we can't do it with the Gradle plugin anyway but we can use:
```sh
gcloud app deploy app.yaml
gcloud app deploy service2/app.yaml
```

**Comment by wkurniawan07** [comment]
> there is usually no need to just deploy the app, can't think of a use case for it

@LiHaoTan The fact is 90-95% of the time `cron.yaml`, `queue.yaml`, `index.yaml` did not change across versions.

If I use deployables, I'm essentially removing the ability to deploy just the app. With a new deployment command I don't have anything to lose with the trade-off of just 3 additional letters to type when deploying. Why are you so against it?

File: docs/development.md [hunk 1]
```diff
@@ -172,6 +173,14 @@ Several configurations are provided by default:
 * `Local tests` - Runs `src/test/testng-local.xml`, all the tests that need to be run locally by developers. `Dev green` means passing all the tests in this configuration.
 * `Failed tests` - Runs `test-output/testng-failed.xml`, which is generated if a test run results in some failures. This will run only the failed tests.
 
+#### GodMode
```
Comment (damithc) [inline docs/development.md]: Wrote an alternative explanation. Use all/part of it as you see fit, as is or in a revised form.

----
To ensure that the UI generated by the app is same as the expected UI, some of our tests compare the raw html text of Web pages generated by the app (let's call them _actual-html_) against the corresponding content of _expected-html_ files previously recorded. This means these _expected-html_ files need to be updated every time you do a UI-related code change (e.g. add a new feature to the UI). Assuming you have already done the UI-related code change and tested in manually, the procedure to update expected-html files of a test is as follows:
1. Run the test in a special mode we call the _God Mode_ (see [godmode.md](godmode.md) for more info) which captures the Web pages being generated by the app and use that new content to overwrites existing expected-html pages.
1. Manually examine the diff of the updated expected-html files to ensure the changes are as expected.
After that, future runs of the test (in normal mode) will be comparing the actual-html against these updated expected-html files that you generated using God Mode. 

Note that the same God Mode technique is used to ensure the content of emails generated by the app are as expected.

**Comment by LiHaoTan** [comment]
@wkurniawan07 I am not very against it, but it seems better to use an in-built solution if it is available. After all, if a project have multiple services, then deployables might have to be modified anyway I think. I don't think there's much performance difference always deploying everything together?

**Pull request merged**